common/src/main/java/com/microsoft/identity/common/components/AndroidPlatformComponentsFactory.java

```diff
@@ -0,0 +1,123 @@
+ */
+public class AndroidPlatformComponentsFactory {
+    /**
+     * True if all of the platform-dependent static classes have been initialized.
+     */
+    private static boolean sGlobalStateInitalized = false;
+
+    /**
+     * Initializes platform-dependent static classes.
+     */
+    public static synchronized void initializeGlobalStates(@NonNull final Context context){
+        if (!sGlobalStateInitalized) {
+            HttpCache.initialize(context);
+            Device.setDeviceMetadata(new AndroidDeviceMetadata());
+            Logger.setAndroidLogger();
+            sGlobalStateInitalized = true;
+        }
+    }
+
+    /**
+     * Creates an {@link IPlatformComponents} object from a {@link Context}.
+     *
+     * @param context an application context.
+     **/
+    public static IPlatformComponents createFromContext(@NonNull final Context context) {
+        return create(context, null, null);
+    }
+
+    /**
+     * Creates an {@link IPlatformComponents} object from an {@link Activity} and, optionally, a {@link Fragment}.
+     *
+     * @param activity an activity where an interactive session will be attached to.
+     * @param fragment a fragment where an interactive session will be attached to.
+     **/
+    public static IPlatformComponents createFromActivity(@NonNull final Activity activity,
+                                                               @Nullable final Fragment fragment) {
+
+        return create(activity.getApplicationContext(), activity, fragment);
+    }
+
+    @SuppressWarnings(WarningType.rawtype_warning)
+    private static IPlatformComponents create(@NonNull final Context context,
+                                              @Nullable final Activity activity,
+                                              @Nullable final Fragment fragment) {
+        initializeGlobalStates(context);
+
+        final PlatformComponents.PlatformComponentsBuilder builder = PlatformComponents.builder();
+        fillBuilder(builder, context, activity, fragment);
+        return builder.build();
+    }
+
+    @SuppressWarnings(WarningType.rawtype_warning)
+    public static void fillBuilder(@NonNull final PlatformComponents.PlatformComponentsBuilder builder,
+                                   @NonNull final Context context,
+                                   @Nullable final Activity activity,
+                                   @Nullable final Fragment fragment) {
+        builder.storageEncryptionManager(new AndroidAuthSdkStorageEncryptionManager(context, null))
+                .clockSkewManager(new AndroidClockSkewManager(context))
+                .broadcaster(new AndroidBroadcaster(context))
+                .popManagerLoader(new AndroidPopManagerLoader(context))
+                .storageLoader(new AndroidStorageLoader(context))
+                .platformUtil(new AndroidPlatformUtil(context, activity))
+                .httpClientWrapper(new DefaultHttpClientWrapper());
```

Comment:
If we invoke one of the create methods in this class such as `createFromActivity` multiple (n) times, then I believe that leads to n number of objects created for each of these i.e. `AndroidBroadcaster`, `AndroidPopManagerLoader` etc.?

Do we want that?

Reply:
I think it's okay _for the scope of the factory_. That's what a factory does. If we want to optimize, we should be optimizing this on the higher level - making sure that we only create it once per flow.

(These classes take a context as parameter, and I'm not sure if having static objects that rely on a certain context object is going to be safe)